Wrong obligation cause span of associated type mismatch

### Bug

The following case ([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=074a0b0c7c96ae12ff9d48104d6e6285)) is slightly modifed based on that in #57663.

```rust
trait Bar {
    type Ok;
    type Sibling: Bar2<Ok=char>; // Here is modified, not using Self::Ok anymore
}
trait Bar2 {
    type Ok;
}

struct Foo;
struct Foo2;

impl Bar for Foo {
    type Ok = ();
    type Sibling = Foo2;
}
impl Bar2 for Foo2 {
    type Ok = u32;
}

fn main() {}
```

Error:

```
error[E0271]: type mismatch resolving `<Foo2 as Bar2>::Ok == char`
  --> src/main.rs:13:15
   |
13 |     type Ok = ();
   |               ^^ expected `u32`, found `char`
```

This is wrong. It points to `<Foo as Bar>::Ok`, which is not related to the error. `<Foo as Bar>::Ok` just happens to have the same identifier as the mismatched associated type `Bar2::Ok`. If I change the identifier in this line to something else ([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c8cc98e279d95bed716967de5a41ac76)), the error changes to the following. It is still not clear, but at least not that wrong.

```
error[E0271]: type mismatch resolving `<Foo2 as Bar2>::Ok == char`
  --> src/main.rs:12:6
   |
12 | impl Bar for Foo {
   |      ^^^ expected `u32`, found `char`
```

### Cause

This is introduced by the PR #65288. The PR tries to make compile errors of associated type mismatch more informative. For an obligation related to a `ty::ProjectionPredicate`, the PR ([these lines](https://github.com/rust-lang/rust/pull/65288/files#diff-a08e8a5bc604df467fdec291a4a38544R224-R233)) iterates over associated types in the current `impl` (`impl Bar for Foo` in this case) and looks for one having the same identifier as the projected item (`Bar2::Ok` in this case). This is wrong and leads to the bug above. It just happens to work for the case in #57663, because in that case `Bar` requires `type Sibling: Bar2<Ok=Self::Ok>` and the two associated types have the same name `Ok`.

PR #69793 refactors this code, adding a new branch ([here](https://github.com/rust-lang/rust/pull/69793/files#diff-236fd1c050048463d34c039bc71b21acR169-R179)) to look for `ProjectionPredicate::ty` instead (`char` in this case, not an associated type). This would work correctly for both this case and #57633. But this PR does not remove the old logic.

### PR
I tried to delete the wrong branch and confirmed that all test cases passed as before. I will create a PR for that change shortly.

--- a/src/librustc_trait_selection/traits/wf.rs
+++ b/src/librustc_trait_selection/traits/wf.rs
@@ -172,25 +172,18 @@ fn extend_cause_with_original_assoc_item_obligation<'tcx>(
         };
     match pred.kind() {
         ty::PredicateKind::Projection(proj) => {
-            // The obligation comes not from the current `impl` nor the `trait` being
-            // implemented, but rather from a "second order" obligation, like in
-            // `src/test/ui/associated-types/point-at-type-on-obligation-failure.rs`.
-            let trait_assoc_item = tcx.associated_item(proj.projection_def_id());
-            if let Some(impl_item_span) =
-                items.iter().find(|item| item.ident == trait_assoc_item.ident).map(fix_span)
-            {
-                cause.span = impl_item_span;
-            } else {
-                let kind = &proj.ty().skip_binder().kind;
-                if let ty::Projection(projection_ty) = kind {
-                    // This happens when an associated type has a projection coming from another
-                    // associated type. See `traits-assoc-type-in-supertrait-bad.rs`.
-                    let trait_assoc_item = tcx.associated_item(projection_ty.item_def_id);
-                    if let Some(impl_item_span) =
-                        items.iter().find(|item| item.ident == trait_assoc_item.ident).map(fix_span)
-                    {
-                        cause.span = impl_item_span;
-                    }
+            // The obligation comes not from the current `impl` nor the `trait` being implemented,
+            // but rather from a "second order" obligation, where an associated type has a
+            // projection coming from another associated type. See
+            // `src/test/ui/associated-types/point-at-type-on-obligation-failure.rs` and
+            // `traits-assoc-type-in-supertrait-bad.rs`.
+            let kind = &proj.ty().skip_binder().kind;
+            if let ty::Projection(projection_ty) = kind {
+                let trait_assoc_item = tcx.associated_item(projection_ty.item_def_id);
+                if let Some(impl_item_span) =
+                    items.iter().find(|item| item.ident == trait_assoc_item.ident).map(fix_span)
+                {
+                    cause.span = impl_item_span;
                 }
             }
         }

--- a/src/test/ui/associated-types/issue-72806.rs
+++ b/src/test/ui/associated-types/issue-72806.rs
@@ -1,0 +1,20 @@
+trait Bar {
+    type Ok;
+    type Sibling: Bar2<Ok=char>;
+}
+trait Bar2 {
+    type Ok;
+}
+
+struct Foo;
+struct Foo2;
+
+impl Bar for Foo {  //~ ERROR type mismatch resolving `<Foo2 as Bar2>::Ok == char`
+    type Ok = ();
+    type Sibling = Foo2;
+}
+impl Bar2 for Foo2 {
+    type Ok = u32;
+}
+
+fn main() {}

--- a/src/test/ui/associated-types/issue-72806.stderr
+++ b/src/test/ui/associated-types/issue-72806.stderr
@@ -1,0 +1,9 @@
+error[E0271]: type mismatch resolving `<Foo2 as Bar2>::Ok == char`
+  --> $DIR/issue-72806.rs:12:6
+   |
+LL | impl Bar for Foo {
+   |      ^^^ expected `u32`, found `char`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0271`.

